File: dotnet/AbirQu/AbirQuGate.cs
using System;
using System.Runtime.InteropServices;

namespace AbirQu
{
    /// <summary>
    /// Gate descriptor for batch circuit execution.
    /// Mirrors the C AbirQuGate struct.
    /// </summary>
    [StructLayout(LayoutKind.Sequential, Pack = 8)]
    public struct AbirQuGate
    {
        public byte GateType;
        public byte Pad0;
        public byte Pad1;
        public byte Pad2;
        public uint Ctrl;
        public uint Target;
        public double Param;

        public AbirQuGate(byte gateType, uint ctrl, uint target, double param)
        {
            GateType = gateType;
            Pad0 = 0;
            Pad1 = 0;
            Pad2 = 0;
            Ctrl = ctrl;
            Target = target;
            Param = param;
        }

        public AbirQuGate(byte gateType, uint qubit) : this(gateType, qubit, 0, 0.0)
        {
        }

        public static AbirQuGate H(uint q) => new AbirQuGate(Simulator.GATE_H, q);
        public static AbirQuGate X(uint q) => new AbirQuGate(Simulator.GATE_X, q);
        public static AbirQuGate Y(uint q) => new AbirQuGate(Simulator.GATE_Y, q);
        public static AbirQuGate Z(uint q) => new AbirQuGate(Simulator.GATE_Z, q);
        public static AbirQuGate CNOT(uint ctrl, uint tgt) => new AbirQuGate(Simulator.GATE_CNOT, ctrl, tgt, 0.0);
        public static AbirQuGate RX(uint q, double angle) => new AbirQuGate(Simulator.GATE_RX, q, q, angle);
        public static AbirQuGate RY(uint q, double angle) => new AbirQuGate(Simulator.GATE_RY, q, q, angle);
        public static AbirQuGate RZ(uint q, double angle) => new AbirQuGate(Simulator.GATE_RZ, q, q, angle);
    }
}

File: swift/AbirQuGate.swift
import Foundation

/// Gate descriptor for batch circuit execution.
/// Mirrors the C AbirQuGate struct.
public struct AbirQuGate {
    public var gateType: UInt8
    public var pad0: UInt8 = 0
    public var pad1: UInt8 = 0
    public var pad2: UInt8 = 0
    public var ctrl: UInt32
    public var target: UInt32
    public var param: Double

    public init(gateType: UInt8, ctrl: UInt32, target: UInt32, param: Double) {
        self.gateType = gateType
        self.ctrl = ctrl
        self.target = target
        self.param = param
    }

    public static func h(_ q: UInt32) -> AbirQuGate {
        return AbirQuGate(gateType: AbirQuSimulator.GATE_H, ctrl: q, target: 0, param: 0.0)
    }

    public static func x(_ q: UInt32) -> AbirQuGate {
        return AbirQuGate(gateType: AbirQuSimulator.GATE_X, ctrl: q, target: 0, param: 0.0)
    }

    public static func y(_ q: UInt32) -> AbirQuGate {
        return AbirQuGate(gateType: AbirQuSimulator.GATE_Y, ctrl: q, target: 0, param: 0.0)
    }

    public static func z(_ q: UInt32) -> AbirQuGate {
        return AbirQuGate(gateType: AbirQuSimulator.GATE_Z, ctrl: q, target: 0, param: 0.0)
    }

    public static func cnot(ctrl: UInt32, tgt: UInt32) -> AbirQuGate {
        return AbirQuGate(gateType: AbirQuSimulator.GATE_CNOT, ctrl: ctrl, target: tgt, param: 0.0)
    }

    public static func rx(_ q: UInt32, angle: Double) -> AbirQuGate {
        return AbirQuGate(gateType: AbirQuSimulator.GATE_RX, ctrl: q, target: q, param: angle)
    }

    public static func ry(_ q: UInt32, angle: Double) -> AbirQuGate {
        return AbirQuGate(gateType: AbirQuSimulator.GATE_RY, ctrl: q, target: q, param: angle)
    }

    public static func rz(_ q: UInt32, angle: Double) -> AbirQuGate {
        return AbirQuGate(gateType: AbirQuSimulator.GATE_RZ, ctrl: q, target: q, param: angle)
    }
}
